### PR TITLE
Rejection Sampling on GSM8k

### PR DESCRIPTION
**Pull request by AlexPiche**
# Rejection Sampling on GSM8k

- [x] Implement a simple reasoning architecture. `COTMathAgent`
- [x] Fix vLLM get log probs OOM #77 
- [x] Do not include empty user message in `nodes.py`
- [x] Launch training jobs with accelerate to handle multi-gpu. #76 (There is an NCCL timeout for large gradient accumulation passes #104)
- [x] Get dataset stats (max, min, mean length) in `finetune.py`
- [x] Fix `finetune` to start from the latest model instead of the base one.
- [x] Move `finetune/rl` helper functions into `finetune/rl/utils.py`
- [x] Implicit KL like Eq3 https://arxiv.org/abs/2402.14740  
- [x] Clean up rl orchestrator: remove discounting and max steps 

## Tape Browser PR from @rizar 

- [x] add a script to launch the tape browser
- [x] save tapes in a single json file, which is a format that the default tape browser can load
- [x] add a script to gather legacy tapes in single file so that we can view legacy tapes in the browser
- [x] load LLM Calls 100 at a time to speed up tape browser loading time (the speedup was huge when the browser did not know where to look for LLM Calls; now that I added the search for llm_calls.sqlite to browser.py, loading is taking some time again)

Examples:
    run tape browser like this: `python -m examples.rl_gsm8k.browse outputs/yolo2_4gpu_lr1e-6/tapes/train/0/all`

## Reasoning Architecture

Screenshot 2024-11-18 at 10 10 54 AM

## Learning curve

`python examples/rl_gsm8k/orchestrate_rl.py finetune.rl.algo=reinforce finetune.train_batch_size=4 finetune.gradient_accumulation_passes=1 finetune.rl.implicit_kl_coef=0.0 finetune.rl.kl_coef=0.0 finetune.rl.use_advantages=false +finetune.rl.relu_weights=true use_rejection_sampling=true test_every_n_iterations=5 finetune.learning_rate=0.000001 finetune.gradient_clipping_threshold=1.0 finetune.save_checkpoint_steps=8 finetune.weight_decay=0.1 max_agent_forks=5000 attempts=8`

Screenshot 2024-11-19 at 9 46 08 AM

# Reproducing GSM8k

Screenshot 2024-11-19 at 4 33 43 PM

### CODE REVIEW

**Comment by rizar** [comment]
will address #77  and #76

File: tapeagents/finetune/rl/__init__.py [hunk 1]
```diff
@@ -196,6 +201,17 @@ def update_advantages(dataset: Dataset, config: RLConfig) -> Dataset:
 
     """
     df = dataset.to_pandas()
+    
+    logging.info("Update Reward with Implicit KL")
+    def calculate_reward_with_implicit_kl(row):
```
Comment (rizar) [inline tapeagents/finetune/rl/__init__.py:206]: this function does not use any variables of the function in which it is located, I'd move it to the top-level. Inner functions are often not so great for readability.

File: requirements.finetune.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-accelerate==0.34.2
+accelerate==1.0.1
```
Comment (rizar) [inline requirements.finetune.txt:1]: @ollmer objections?

File: conf/rl_gsm8k.yaml [hunk 3]
```diff
@@ -31,12 +34,12 @@ finetune:
 vllm_config:
   vllm_kwargs:
     --download-dir: /mnt/llmd/base_models/ 
-    --max-model-len: 8000
     --gpu-memory-utilization: 0.9
     # VLLM get log probs OOM https://github.com/vllm-project/vllm/issues/5907
     --enable-chunked-prefill: ""
 
 output_dir: outputs/rl_gsm8k
+accelerate_cfg_path: conf/deepspeed/accelerate_base.yaml
```
Comment (rizar) [inline conf/rl_gsm8k.yaml:42]: we should change the name of this config directory to `conf/accelerate`

**Review by rizar — APPROVED** [review]
looks good! let me go test this now

File: examples/rl_gsm8k/orchestrate_rl.py [hunk 4]
```diff
@@ -309,8 +328,10 @@ def main(cfg: DictConfig):
             test_agent = MathAgent.create(llm=test_llm)
 
             datasets = [("train", train_agent, train_tapes)]
-            if state["iteration"] % cfg.test_every_n_iterations == 0:
-                datasets.append(("test", test_agent, test_tapes))
+            if state["iteration"] % cfg.test_every_n_iterations == 0 and cfg.test_every_n_iterations > 0:
```
Comment (rizar) [inline examples/rl_gsm8k/orchestrate_rl.py:331]: remove before merging :)

File: tapeagents/finetune/checkpoints.py [hunk 5]
```diff
@@ -112,8 +112,6 @@ def load_model(args, model_class, current_dir, is_rl=False):
 
         # when LoRA enabled, always preload the original model, the lora weights will be loaded later
         model_to_load = args.config_name if args.lora.enabled else str(current_dir)
-        if is_rl:
```
Comment (rizar) [inline tapeagents/finetune/checkpoints.py:115]: was it a debug hack?

File: tapeagents/finetune/finetune.py [hunk 6]
```diff
@@ -155,6 +156,10 @@ def run_finetuning_loop(
         logger.info("LR after loading training state: %.2E" % training_metrics.lr)
         dt = log_time(dt, "finetune/training_state_load")
 
+    if accelerator.is_main_process:
```
Comment (rizar) [inline tapeagents/finetune/finetune.py:159]: You have just made it mandatory to make a full pass over an arbitrarily large training set.... not a great default behavior.

Can you implement this as a running metric in the training loop instead? You can report it every time the gradient accumulation is finished.

File: tapeagents/finetune/rl/__init__.py [hunk 7]
```diff
@@ -128,6 +107,7 @@ def rl_step(model, batch, config: RLConfig) -> tuple[torch.Tensor, dict[str, flo
     log_ratio_new_old = new_log_probs - old_logprobs
     ratio_new_old = torch.exp(log_ratio_new_old)
     weights = advantages if config.use_advantages else rewards
+    weights = torch.clamp(weights, min=0) if config.relu_weights else weights
```
Comment (rizar) [inline tapeagents/finetune/rl/__init__.py:110]: where does this trick come from? can you put a paper reference? where do we use it?

File: tapeagents/finetune/rl/__init__.py [hunk 8]
```diff
@@ -197,6 +174,13 @@ def update_advantages(dataset: Dataset, config: RLConfig) -> Dataset:
     """
     df = dataset.to_pandas()
 
+    if config.implicit_kl_coef > 0:
```
Comment (rizar) [inline tapeagents/finetune/rl/__init__.py:177]: This adds negative KL to reward, I suppose. Can we call it `config.reward_minus_kl`?

File: tapeagents/observe.py [hunk 9]
```diff
@@ -15,7 +15,7 @@
 logger = logging.getLogger(__name__)
 
 _checked_sqlite = False
-_ACTIVE_MANAGER: Optional["SQLiteQueueManager"] = None
+_ACTIVE_MANAGER: Optional["SQLiteWriterThread"] = None
```
Comment (rizar) [inline tapeagents/observe.py:18]:
```suggestion
_WRITER_THREAD: Optional["SQLiteWriterThread"] = None
```

**Review by rizar — CHANGES_REQUESTED** [review]
Looks mostly good! We should not traverse the whole training set every time the script is launched though. 

Note for the re-review:
- check ReLU weights citation 
- check running aggregation of input lengths

File: requirements.finetune.txt (hunk 2 above)
Comment (ollmer) [reply requirements.finetune.txt:1]: No

**Review by rizar — APPROVED** [review]
lgtm!
